pennylane/qchem/vibrational/bosonic_mapping.py

```diff
@@ -0,0 +1,160 @@
+
+    The mapping procedure is described in `arXiv:1507.03271 <https://arxiv.org/pdf/1507.03271>`_.
+
+    Args:
+        bose_operator(BoseWord, BoseSentence): the bosonic operator
+        nstates(int): maximum number of states a boson can occupy
+        ps (bool): whether to return the result as a PauliSentence instead of an
+            operator. Defaults to False.
+        wire_map (dict): a dictionary defining how to map the states of
+            the Bose operator to qubit wires. If None, integers used to
+            label the bosonic states will be used as wire labels. Defaults to None.
+        tol (float): tolerance for discarding the imaginary part of the coefficients
+
+    Returns:
+        a linear combination of qubit operators
+
+    **Example**
+
+    >>> w = qml.qchem.BoseWord({(0, 0): "+"})
+    >>> qml.qchem.binary_mapping(w, nstates=4)
+    0.6830127018922193 * X(0)
+    + -0.1830127018922193 * X(0) @ Z(1)
+    + -0.6830127018922193j * Y(0)
+    + 0.1830127018922193j * Y(0) @ Z(1)
+    + 0.3535533905932738 * X(0) @ X(1)
+    + -0.3535533905932738j * X(0) @ Y(1)
+    + 0.3535533905932738j * Y(0) @ X(1)
+    + (0.3535533905932738+0j) * Y(0) @ Y(1)
+    """
+
+    qubit_operator = _binary_mapping_dispatch(bose_operator, nstates, tol=tol)
+
+    wires = list(bose_operator.wires) or [0]
+    identity_wire = wires[0]
+    if not ps:
+        qubit_operator = qubit_operator.operation(wire_order=[identity_wire])
+
+    if wire_map:
+        return qubit_operator.map_wires(wire_map)
+
+    return qubit_operator
+
+
+@singledispatch
+def _binary_mapping_dispatch(bose_operator, nstates, tol):
+    """Dispatches to appropriate function if bose_operator is a BoseWord or BoseSentence."""
+    raise ValueError(f"bose_operator must be a BoseWord or BoseSentence, got: {bose_operator}")
+
+
+@_binary_mapping_dispatch.register
+def _(bose_operator: BoseWord, nstates, tol=None):
+
+    if nstates < 2:
+        raise ValueError(
+            f"Number of states a boson can occupy cannot be less than 2, provided {nstates}."
+        )
+    nqub_per_boson = int(np.ceil(np.log2(nstates)))
+
+    cr_op = np.zeros((nstates, nstates))
+    for s in range(nstates - 1):
+        cr_op[s + 1, s] = np.sqrt(s + 1.0)
+
+    op_mat = {"+": cr_op, "-": cr_op.T}
+
+    qubit_operator = PauliSentence({PauliWord({}): 1.0})
+
+    for item in bose_operator.items():
+        (_, b_idx), sign = item
+        op = PauliSentence()
+        sparse_opmat = np.nonzero(op_mat[sign])
+        for i, j in zip(*sparse_opmat):
+            coeff = op_mat[sign][i][j]
+
+            binary_row = list(map(int, bin(i)[2:]))[::-1]
+            if nqub_per_boson > len(binary_row):
+                binary_row += [0] * (nqub_per_boson - len(binary_row))
+
+            binary_col = list(map(int, bin(j)[2:]))[::-1]
+            if nqub_per_boson > len(binary_col):
+                binary_col += [0] * (nqub_per_boson - len(binary_col))
+
```

Comment:
I wonder if there is a better name than `binary_row`, `binary_col`. To clarify, these are the bit strings right? E.g., 
row = [0, 0] == |00>
col = [1, 0]  == <10|

|00><10|
We're supposed to interpret this as
`|0><1| @ |0><0|`and do the mapping with `_get_pauli_op`, correct?

Reply:
For further clarification, if I have some operator, `b0`, depending on the # of states, this means that it can be interpreted as |0><1|, |1><2|, |2><3|, etc... So in the case of n=3, we have |0><1| or |1><2|, so we do |0><1| @ |1><2| -> |00><01| @ |01|><10|. I did the math manually and it seems to check out, but I'm just trying to clarify my understanding here.

Reply:
Yes these are the bit strings, since we are converting the numbers to binary bit strings, that is why I chose the variable name.